src/coap_tinydtls.c

```diff
@@ -562,11 +698,638 @@ coap_get_tls_library_version(void) {
+      /* SEQUENCE or SEQUENCE OF */
+      tag_data = asn1_abstract_tag(acp, len, ltag, callback);
+      if (tag_data)
+        return tag_data;
+    }
+    acp += len;
+    tlen -= len;
+    tag = asn1_tag_c(&acp, &constructed, &class);
+    len = asn1_len(&acp);
+  }
+  return NULL;
+}
+
+static int
+asn1_derive_keys(coap_tiny_context_t *t_context,
+                 const uint8_t *priv_data, size_t priv_len,
+                 const uint8_t *pub_data, size_t pub_len,
+                 int is_pkcs8)
+{
+  coap_binary_t *test;
+
+  t_context->priv_key = asn1_abstract_tag(priv_data, priv_len,
+                                              0x04, asn1_callback_privkey);
+  if (!t_context->priv_key) {
+    coap_log(LOG_INFO, "EC Private Key (RPK) invalid\n");
+    return 0;
+  }
+
+  if (!is_pkcs8) {
+    /* pkcs8 abstraction tested for valid eliptic curve */
+    test = asn1_abstract_tag(priv_data, priv_len,
+                                    0x06, asn1_callback_curve);
+    if (!test) {
+      coap_log(LOG_INFO, "EC Private Key (RPK) invalid elliptic curve\n");
+      coap_delete_binary(t_context->priv_key);
+      t_context->priv_key = NULL;
+      return 0;
+    }
+    coap_delete_binary(test);
+  }
+
+  t_context->pub_key = asn1_abstract_tag(pub_data, pub_len,
+                                              0x03, asn1_callback_pubkey);
+  if (!t_context->pub_key) {
+    coap_log(LOG_INFO, "EC Public Key (RPK) invalid\n");
+    coap_delete_binary(t_context->priv_key);
+    t_context->priv_key = NULL;
+    return 0;
+  }
+  dtls_set_handler(t_context->d_context, &ec_cb);
+  return 1;
+}
+
+static coap_binary_t *
+ec_abstract_pkcs8_asn1(const uint8_t *asn1_ptr, size_t asn1_length)
+{
+  coap_binary_t *test;
+
+  test = asn1_abstract_tag(asn1_ptr, asn1_length,
+                               0x02, asn1_callback_pkcs8_version);
+  if (!test)
+    return 0;
+
+  coap_delete_binary(test);
+
+  test = asn1_abstract_tag(asn1_ptr, asn1_length,
+                               0x06, asn1_callback_ec_identifier);
+  if (!test)
+    return 0;
+  coap_delete_binary(test);
+
+  test = asn1_abstract_tag(asn1_ptr, asn1_length,
+                               0x06, asn1_callback_curve);
+  if (!test) {
+    coap_log(LOG_INFO, "EC Private Key (RPK) invalid elliptic curve\n");
+    return 0;
+  }
+  coap_delete_binary(test);
+
+  test = asn1_abstract_tag(asn1_ptr, asn1_length,
+                               0x04, asn1_callback_ec_key);
+  return test;
+}
+
+static coap_binary_t *
+pem_decode_mem_asn1(const char *begstr, const uint8_t *str)
+{
+  char *bcp = str ? strstr((const char*)str, begstr) : NULL;
+  char *tcp = bcp ? strstr(bcp, "-----END ") : NULL;
+
+  if (bcp && tcp) {
+    bcp += strlen(begstr);
+    return pem_base64_decode ((const uint8_t *)bcp, tcp - bcp);
+  }
+  return NULL;
+}
+
+#include <stdio.h>
+#include <sys/stat.h>
+
+static uint8_t *read_file_mem(const char* file, size_t *length) {
+  FILE *f = fopen(file, "r");
```

Comment:
These POSIX file operations would be a problem on some embedded OSes that often do not have a full-fledged file system. (Not sure if we have to do anything about this in this particular PR; I suppose we simply do not allow COAP_PKI_KEY_PEM as key_type in setup_data and make this set of functions conditional.)

Reply:
An alternative way would be to disable COAP_PKI_KEY_PEM support, but make coap-server/coap-client convert the files to a memory load and use COAP_PKI_KEY_PEM_BUF instead.  Relatively straight forward.

What do you think?

Reply:
Sounds good.

Reply:
Working on it.

Reply:
Code pushed as a separate commit for ease of review.  Commits can be merged if required.